### PR TITLE
Turn on -Wsign-compare

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74996

This is enabled on some of our internal builds, is a common source
of fbcode only errors and apparently we are relatively clean on it.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

**Review by dagitses — COMMENTED** [review]
looks good to me, but deferring to nikita for cmake changes.

File: c10/util/int128.cpp [hunk 1]
```diff
@@ -171,7 +171,7 @@ std::ostream& operator<<(std::ostream& o, const uint128& b) {
 
   // Add the requisite padding.
   std::streamsize width = o.width(0);
-  if (width > rep.size()) {
+  if (static_cast<std::string::size_type>(width) > rep.size()) {
```
[the remark this answers is not in the document]
Comment (dagitses) [reply c10/util/int128.cpp:174]: thanks, that's good enough for me!

**Review by malfet — APPROVED** [review]
Looks good to me, but I wonder whether this is already covered by https://github.com/pytorch/pytorch/pull/75085 
Also, why enable it just for C10 instead of the entire codebase?

**Comment by ezyang** [comment]
@malfet At least one difference is that this change applies to gcc, while yours applies to clang only. I didn't enable for the whole codebase because I noticed there were a lot of existing failures and I didn't want to fix them all immediately.

**Comment by malfet** [comment]
@ezyang I believe I've fixed all those violations(see stack ending with https://github.com/pytorch/pytorch/pull/75084), so feel free to try enabling it globally. Also, I wonder if `-Wsign-compare` should be enabled instead of its `-Werror` variant  and then respect WERROR option, which is enabled for all CI/CD builds right now.
Very likely, users with newer version of compiler will hit some violation and will have no way of suppressing those

**Comment by ezyang** [comment]
>  Also, I wonder if -Wsign-compare should be enabled instead of its -Werror variant and then respect WERROR option, which is enabled for all CI/CD builds right now.

Hmm, I'm not sure what you mean here. How would this help with the new compiler case?

**Pull request closed without merging**